py/dml/types_test.py

```diff
@@ -0,0 +1,282 @@
+                    TLayout("big-endian", {}),
+                    TDevice("a")):
+            typ_clone = typ.clone()
+            self.assertTrue(
+                realtype(typ_clone).eq(realtype(typ)))
+            self.assertTrue(
+                realtype(typ).eq(realtype(typ_clone)))
+            typ_clone.const = True
+            self.assertFalse(typ.const)
+            typ = typ_clone.clone()
+            self.assertTrue(typ.const)
+
+        # special case for TraitList, because realtype requires global
+        # state (typedefs)
+        typ = TTraitList("a")
+        typ_clone = typ.clone()
+        self.assertTrue(typ.eq(typ_clone))
+        self.assertTrue(typ_clone.eq(typ))
+        dml.globals.dml_version = (1, 2)
+        # types which do not support clone
+        with self.assertRaises(logging.ICE):
+            TUnknown().clone()
+
+class Typ1(DMLType):
+    def describe(self):
+        return 'Typ1'
+    def clone(self):
+        return Typ1(self.const)
+
+class Typ2(DMLType):
+    def describe(self):
+        return 'Typ2'
+    def clone(self):
+        return Typ1(self.const)
+
+class TestEq(unittest.TestCase):
+    def assert_eq(self, t1, t2):
+        self.assertTrue(t1.eq(t2))
+        self.assertTrue(t2.eq(t1))
+        self.assertEqual(t1.hashed(), t2.hashed())
+        try:
+            self.assertEqual(t1.key(), t2.key())
+        except DMLUnkeyableType:
+            pass
+
+    def assert_neq(self, t1, t2, hash_collision=False):
+        self.assertFalse(t1.eq(t2))
+        self.assertFalse(t2.eq(t1))
+        if not hash_collision:
+            self.assertNotEqual(t1.hashed(), t2.hashed())
+            try:
+                self.assertNotEqual(t1.key(), t2.key())
+            except DMLUnkeyableType:
+                pass
+
+    def test_DMLType(self):
+        self.assert_eq(Typ1(), Typ1())
+        self.assert_neq(Typ1(True), Typ1())
+        self.assert_neq(Typ1(), Typ2())
+
+    def test_IntegerType(self):
+        self.assert_eq(TInt(8, False), TInt(8, False))
+        self.assert_neq(TInt(16, False), TInt(8, False))
+        # Not equivalent even though the C representations of uint16 and uint13
+        # are compatible
+        self.assert_neq(TInt(16, False), TInt(13, False))
+        # Signedness
+        self.assert_neq(TInt(8, True), TInt(8, False))
+
+        # bitfields
+        self.assert_neq(TInt(8, False), TInt(8, False, members={}))
+
+        def bitfields():
+            return TInt(32, False, { 'a': (TInt(8, False), 13, 6),
+                                     'b': (TInt(13, False), 27, 15) })
+
+        self.assert_eq(TInt(8, False, members={}), TInt(8, False, members={}))
+        self.assert_eq(bitfields(), bitfields())
+
+        t = bitfields()
+
+        def neq_bitfields_testcase():
+            nonlocal t
+            self.assert_neq(bitfields(), t)
+            t = bitfields()
+
+        # Presence of members matter
+        del t.members['a']
+        neq_bitfields_testcase()
```

Comment:
macro-like function, probably clearer to just expand it every time. AFAICS, the cost in code size is zero

Reply:
The reason I did this here is because I wanted to reduce the boilerplate and noise needed for each test-case specification. The fact this pattern incurs noise _before_ the mass of specifications, or the code size would be identical were the function to be expanded are of less relevance to me compared to the fact that each test case would always need two lines of boilerplate in the thick of it instead of one. I feel that hampers readability.

Reply:
My problem with the function is that it is not an abstraction: it does capture a recurring sequence of operations, but the best description I can come up with is "do the last thing we usually do in the end of one testcase, and then do the first thing we usually do in the beginning of the next". Every time you read a call to the function, you need to do this expansion in your head to understand what will happen. This makes the reading heavier than just reading those two lines of code straight up.

Reply:
In that case, how about:

```

        @contextmanager
        def neq_bitfields_testcase():
            t = bitfields()
            yield t
            self.assert_neq(bitfields(), t)
```

This _is_ actually an abstraction, and would actually resemble the way I'd do it in a **better** language like Haskell (I'd do a higher-order function, and then leverage lambdas that have the luxury of actually being able to stretch multiple lines.)

Reply:
We've agreed on the context manager solution